ci-operator/step-registry/cucushift/upgrade/rollback/cucushift-upgrade-rollback-commands.sh

```diff
@@ -0,0 +1,159 @@
+        echo -e "Describing abnormal mcp...\n"
+        oc get mcp --no-headers | awk '$3 != "True" || $4 != "False" || $5 != "False" {print $1}' | while read mcp; do echo -e "\n#####oc describe mcp ${mcp}#####\n$(oc describe mcp ${mcp})"; done
+    fi
+}
+
+# Generate the Junit for upgrade
+function createUpgradeJunit() {
+    echo "Generating the Junit for upgrade"
+    if (( FRC == 0 )); then
+      cat >"${ARTIFACT_DIR}/junit_upgrade.xml" <<EOF
+<?xml version="1.0" encoding="UTF-8"?>
+<testsuite name="cluster upgrade" tests="1" failures="0">
+  <testcase classname="cluster upgrade" name="upgrade should succeed"/>
+</testsuite>
+EOF
+    else
+      cat >"${ARTIFACT_DIR}/junit_upgrade.xml" <<EOF
+<?xml version="1.0" encoding="UTF-8"?>
+<testsuite name="cluster upgrade" tests="1" failures="1">
+  <testcase classname="cluster upgrade" name="upgrade should succeed">
+    <failure message="">openshift cluster upgrade failed</failure>
+  </testcase>
+</testsuite>
+EOF
+    fi
+}
+
+# Extract oc binary which is supposed to be identical with target release
+function extract_oc(){
+    echo -e "Extracting oc\n"
+    local retry=5 tmp_oc="/tmp/client-2"
+    mkdir -p ${tmp_oc}
+    while ! (env "NO_PROXY=*" "no_proxy=*" oc adm release extract -a "${CLUSTER_PROFILE_DIR}/pull-secret" --command=oc --to=${tmp_oc} ${TARGET});
+    do
+        echo >&2 "Failed to extract oc binary, retry..."
+        (( retry -= 1 ))
+        if (( retry < 0 )); then return 1; fi
+        sleep 60
+    done
+    mv ${tmp_oc}/oc ${OC_DIR} -f
+    export PATH="$PATH"
+    which oc
+    oc version --client
+    return 0
+}
+
+function run_command() {
+    local CMD="$1"
+    echo "Running command: ${CMD}"
+    eval "${CMD}"
+}
+
+# Rollback the cluster to target release
+function rollback() {
+    run_command "oc adm upgrade rollback"
+    echo "Rolling back cluster gets started..."
+}
+
+# Monitor the rollback status
+function check_rollback_status() {
+    local wait_rollback="${TIMEOUT}" out avail progress cluster_version
+    cluster_version="${TARGET_VERSION}"
+    echo "Starting the rollback checking on $(date "+%F %T")"
+    while (( wait_rollback > 0 )); do
+        sleep 5m
+        wait_rollback=$(( wait_rollback - 5 ))
+        if ! ( run_command "oc get clusterversion" ); then
+            continue
+        fi
+        if ! out="$(oc get clusterversion --no-headers || false)"; then
+            echo "Error occurred when getting clusterversion"
+            continue 
+        fi
+        avail="$(echo "${out}" | awk '{print $3}')"
+        progress="$(echo "${out}" | awk '{print $4}')"
+        if [[ ${avail} == "True" && ${progress} == "False" && ${out} == *"Cluster version is ${cluster_version}" ]]; then
+            echo -e "Rollback succeed on $(date "+%F %T")\n\n"
+            return 0
+        fi
```

Comment:
so we're sleeping for 5 minutes before we do anything else, so rollback will have enough time to start, and report Progressing=True? I seen `cluster_version="${TARGET_VERSION}"` few lines above, so in case of a rollback, this is the original version right?
asking because I recall a situation where an upgrade automation failed because Progressing went temporarily to false,  and back to true where upgrade is not yet fully complete, however the script already exited

Reply:
> so in case of a rollback, this is the original version
Right

>asking because I recall a situation where an upgrade automation failed because Progressing went temporarily to false

Good point. So we not only check `progressing`, but also check the `version` and `available`. Does it sound safe enough for you?

Reply:
its ok I believe the combination of those three should be enough